kymatio/scattering1d/frontend/base_frontend.py

```diff
@@ -368,4 +435,761 @@ def _document(cls):
+        else:
+            self.resample_psi_fr = self.resample_phi_fr = resample_filters_fr
+        self.max_pad_factor_fr = max_pad_factor_fr
+        self.out_3D = out_3D
+        self.out_type = out_type
+
+    def build(self):
+        """Check args and instantiate `_FrequencyScatteringBase` object
+        (which builds filters).
+
+        Certain internal data, such as the amount of padding and the wavelet
+        filters to be used in the scattering transform, need to be computed
+        from the parameters given during construction. This function is called
+        automatically during object creation and no subsequent calls are
+        therefore needed.
+        """
+        # don't allow untested (and likely unneeded) combination
+        if not self.aligned and not self.out_3D:
+            raise ValueError("`aligned=False` is only allowed with `out_3D=True`")
+
+        # if config yields no second order coeffs, we cannot do joint scattering
+        if self._no_second_order_filters:
+            raise ValueError("configuration yields no second-order filters; "
+                             "try increasing `J`")
+
+        self._shape_fr = self.get_shape_fr()
+        max_order_fr = 1
+        # number of psi1 filters
+        self._n_psi1 = len(self.psi1_f)
+
+        self.sc_freq = _FrequencyScatteringBase(
+            self._shape_fr, self.J_fr, self.Q_fr, self.F, max_order_fr,
+            self.average_fr, self.aligned, self.oversampling_fr,
+            self.resample_psi_fr, self.resample_phi_fr, self.vectorize,
+            self.out_type, self.out_3D, self.max_pad_factor_fr, self._n_psi1,
+            self.backend)
+        self.finish_creating_filters()
+
+        # detach __init__ args, instead access `sc_freq`'s via `__getattr__`
+        # this is so that changes in attributes are reflected here
+        init_args = ('J_fr', 'Q_fr', 'F', 'average_fr', 'oversampling_fr',
+                     'resample_psi_fr', 'resample_phi_fr', 'max_pad_factor_fr',
+                     'out_3D')
+        for init_arg in init_args:
+            delattr(self, init_arg)
+
+    def get_shape_fr(self):
+        """This is equivalent to `len(x)` along frequency, which varies across
+        `psi2`, so we compute for each.
+        """
+        shape_fr = []
+        for n2 in range(len(self.psi2_f)):
+            j2 = self.psi2_f[n2]['j']
+            max_freq_nrows = 0
+            for n1 in range(len(self.psi1_f)):
+                if j2 > self.psi1_f[n1]['j']:
+                    max_freq_nrows += 1
+            shape_fr.append(max_freq_nrows)
+        return shape_fr
+
+    def finish_creating_filters(self):
+        """Handles necessary adjustments in time scattering filters unaccounted
+        for in default construction.
+        """
+        # ensure phi is subsampled up to (log2_T - 1) for `phi_t * psi_f` pairs
+        max_sub_phi = lambda: max(k for k in self.phi_f if isinstance(k, int))
```

Comment:
This and the way it is used below is too implicit and squishy for my taste

Reply:
It is the shortest and spares duplication. I'm open to alternatives.